### PR TITLE
tests/kernel/sched/schedule_api: don't break the test if not divisible

### LINKED ISSUE #44887: it8xxx2_evb: tests/kernel/sched/schedule_api/ assertion fail
**Describe the bug**
Can not run the test case of tests/kernel/sched/schedule_api/ on it8xxx2_evb board.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? -> it8xxx2_evb 
 - What have you tried to diagnose or workaround this issue?
The slice_ticks is equal to 1639 (t = 200, to_hz = 8192, off = 999, from_hz = 1000, (200x8192+999)/1000 = 1639) , but the assertion shows it should be an even number. 

**To Reproduce**
1. west build -p always -b it8xxx2_evb tests/kernel/sched/schedule_api/
2. echo "intel@123" | sudo -S ~/itedlb4-linux-v104/ite -f build/zephyr/zephyr.bin -s check

**Expected behavior**
The test should pass

**Impact**
The test case fails in the daily test.

**Logs and console output**
```
                START - test_slice_reset
                ASSERTION FAIL [slice_ticks % 2 == 0] @ WEST_TOPDIR/zephyr/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:146
                        timeslice in ticks much be divisible by two
                E:      a0: 00000004    t0: aaaaaaaa
                E:      a1: 00000092    t1: 10000000
                E:      a2: 00f02705    t2: aaaaaaaa
                E:      a3: 00f02700    t3: 00000020
                E:      a4: 00000020    t4: 00000001
                E:      a5: 00001880    t5: aaaaaaaa
                E:      a6: a0000000    t6: aaaaaaaa
                E:      a7: 00000000
                E:      ra: 80001f26
                E:    mepc: 80002e8e
                E: mstatus: 00001880
                E:
                E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
                E: Current thread: 0x80102508 (test_slice_reset)
                Caught system error -- reason 4 0
                Fatal error was unexpected, aborting...
```

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Toolchain: zephyr-sdk-0.13.2
 - Commit SHA or Version used: 6e8cb0d1553a6ffde52f4e047c37c1cbfe23bd46

**Additional context**
None

### PR DESCRIPTION
**Pull request by RuibinChang**
When test tests/kernel/sched/schedule_api, it shows "ASSERTION FAIL:
timeslice in ticks much be divisible by two", then break and fail
the test.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/44887.

On board it8xxx2_evb, it can pass schedule_api test without
the assertion, so I add the floating part back to half_slice_cyc
when the timeslice in ticks can't be divisible by two.
After change, the time slice will be:
1.slice_ticks = (200x8192+999)/1000 = 1639 (not changed)
2.before add the deviation (not handle the floating part):
half_slice_cyc = (1639/2) * (32768/8192) = (819) * (32768/8192) = 3276,
after add the deviation:
half_slice_cyc = 3276 + (32768/8192/2) = 3278,
and it's equal (1639/2) * (32768/8192) = 3278.

Verified by test pattern:
west build -p always -b it8xxx2_evb tests/kernel/sched/schedule_api

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

### CODE REVIEW

**Review by ceolin — COMMENTED** [review]
Is it a particular issue for this platform ? There already  is an assert bellow checking if `slice_ticks` is even.

**Comment by RuibinChang** [comment]
> Is it a particular issue for this platform ? There already is an assert bellow checking if `slice_ticks` is even.

According the issue #44887 description, it happens on board mec172xevb_assy6906 and we also find this happens on it8xxx2_evb.

On board it8xxx2_evb, it can pass schedule_api test without the assertion, shall I remove the assert instead of adding one to slice_ticks? 

Time slice without the assertion:
slice_ticks = (200x8192+999)/1000 = 1639
half_slice_cyc = (1639/2) * (32768/8192) = 3278, but we don't handle the floating part in code => half_slice_cyc = (819) * (32768/8192) = 3276 

3278 - 3276 = 2 HW deviation cycle (= 61us, base on 32768 HW clock), so we will early 61us break the half_slice_cyc 100ms while loop .

**Comment by carlescufi** [comment]
@andyross could you please give input on this PR?

**Comment by RuibinChang** [comment]
Update commit message

File: tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c [hunk 1]
```diff
@@ -143,8 +143,12 @@ void test_slice_reset(void)
 	uint32_t slice_ticks = k_ms_to_ticks_ceil32(SLICE_SIZE);
 	uint32_t half_slice_cyc = k_ticks_to_cyc_ceil32(slice_ticks / 2);
 
-	__ASSERT(slice_ticks % 2 == 0,
-		 "timeslice in ticks much be divisible by two");
+	if (slice_ticks % 2 != 0) {
```
Comment (andyross) [inline tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:146]: Yeah, let's not do this.  The "divisible by two" rule here is a workaround for what amounts to a bug in the test (it's making assumptions about time slice size but doing the math in a foreign unit).  The test logic is only going to work reliably if you follow this weird rule, which is why that assertion exists.

Can we either fix the test to do its math in tick units or change the platform config to not trigger the edge case?

**Review by andyross — CHANGES_REQUESTED** [review]

File: tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c (hunk 1 above)
Comment (RuibinChang) [reply tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:146]: Done, half_slice_cyc is added the floating part back that was omitted, when timeslice can't be divisible by two.

**Comment by hunterhu3000** [comment]
The new modifications can also make the test pass. But it replaced the assert with the if condition.
@andyross  Is it OK to you?

**Comment by hunterhu3000** [comment]
The new modifications can also make the test pass. But it replaced the assert with the if condition.
@andyross  Is it OK to you?

**Review by andyross — APPROVED** [review]
This seems more reasonable.  I'd still prefer we fix the math in the test, but at least this way we're not exposing other platforms to weird failures.

**Comment by RuibinChang** [comment]
Any comment?
If not, one more approval will be appreciate.

**Review by dcpleung — APPROVED** [review]
